### ISSUE TITLE
[SR-13728] @objc optional field can't be written in protocol extension method

### ISSUE THREAD

**Issue by kateinoigakukun** [apple/swift#56125]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-13728      |
   |Radar             | rdar://problem/70298051         |
   |Original Reporter | @kateinoigakukun      |
   |Type              | Bug    |

<details>
  <summary>Environment</summary>

Toolchain: swift-DEVELOPMENT-SNAPSHOT-2020-10-07-a

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 471668fec747083c6c83f4b5428a082d

  </details>

**Issue Description:**

This code can't be compiled with swift-DEVELOPMENT-SNAPSHOT-2020-10-07-a

``` java
import Foundation
@objc
public protocol C {
    @objc optional
    var field1: ObjCBool { get set }
}

extension C {
    public func set(value: Bool) {
        self[keyPath: \C.field1] = ObjCBool(value)
    }
}
```

Error output:

``` java
SIL verification failed: keypath value type should match value type of keypath pattern
  Optional<ObjCBool>
  ObjCBool
Verifying instruction:
->   %6 = keypath $ReferenceWritableKeyPath<C, Optional<ObjCBool>>, (objc "field1"; root $C; settable_property $ObjCBool,  id #C.field1!getter.foreign : <Self where Self : C> (Self) -> () -> ObjCBool, getter @$s4main1CP6field110ObjectiveC8ObjCBoolVvpAaB_pTK : $@convention(thin) (@in_guaranteed C) -> @out ObjCBool, setter @$s4main1CP6field110ObjectiveC8ObjCBoolVvpAaB_pTk : $@convention(thin) (@in_guaranteed ObjCBool, @in_guaranteed C) -> ()) // users: %20, %16
     %16 = apply %15<C, Optional<ObjCBool>>(%11, %6, %13) : $@convention(thin) <τ_0_0, τ_0_1> (@in_guaranteed τ_0_0, @guaranteed ReferenceWritableKeyPath<τ_0_0, τ_0_1>, @in τ_0_1) -> ()
     destroy_value %6 : $ReferenceWritableKeyPath<C, Optional<ObjCBool>> // id: %20
In function:
// C.set(value:)
sil [ossa] @$s4main1CPAAE3set5valueySb_tF : $@convention(method) <Self where Self : C> (Bool, @guaranteed Self) -> () {
// %0 "value"                                     // users: %9, %2
// %1 "self"                                      // users: %4, %3
bb0(%0 : $Bool, %1 : @guaranteed $Self):
  debug_value %0 : $Bool, let, name "value", argno 1 // id: %2
  debug_value %1 : $Self, let, name "self", argno 2 // id: %3
  %4 = copy_value %1 : $Self                      // user: %5
  %5 = init_existential_ref %4 : $Self : $Self, $C // user: %12
  %6 = keypath $ReferenceWritableKeyPath<C, Optional<ObjCBool>>, (objc "field1"; root $C; settable_property $ObjCBool,  id #C.field1!getter.foreign : <Self where Self : C> (Self) -> () -> ObjCBool, getter @$s4main1CP6field110ObjectiveC8ObjCBoolVvpAaB_pTK : $@convention(thin) (@in_guaranteed C) -> @out ObjCBool, setter @$s4main1CP6field110ObjectiveC8ObjCBoolVvpAaB_pTk : $@convention(thin) (@in_guaranteed ObjCBool, @in_guaranteed C) -> ()) // users: %20, %16
  %7 = metatype $@thin ObjCBool.Type              // user: %9
  // function_ref ObjCBool.init(_:)
  %8 = function_ref @$s10ObjectiveC8ObjCBoolVyACSbcfC : $@convention(method) (Bool, @thin ObjCBool.Type) -> ObjCBool // user: %9
  %9 = apply %8(%0, %7) : $@convention(method) (Bool, @thin ObjCBool.Type) -> ObjCBool // user: %10
  %10 = enum $Optional<ObjCBool>, #Optional.some!enumelt, %9 : $ObjCBool // user: %14
  %11 = alloc_stack $C                            // users: %19, %18, %16, %12
  store %5 to [init] %11 : $*C                    // id: %12
  %13 = alloc_stack $Optional<ObjCBool>           // users: %17, %16, %14
  store %10 to [trivial] %13 : $*Optional<ObjCBool> // id: %14
  // function_ref swift_setAtReferenceWritableKeyPath
  %15 = function_ref @swift_setAtReferenceWritableKeyPath : $@convention(thin) <τ_0_0, τ_0_1> (@in_guaranteed τ_0_0, @guaranteed ReferenceWritableKeyPath<τ_0_0, τ_0_1>, @in τ_0_1) -> () // user: %16
  %16 = apply %15<C, Optional<ObjCBool>>(%11, %6, %13) : $@convention(thin) <τ_0_0, τ_0_1> (@in_guaranteed τ_0_0, @guaranteed ReferenceWritableKeyPath<τ_0_0, τ_0_1>, @in τ_0_1) -> ()
  dealloc_stack %13 : $*Optional<ObjCBool>        // id: %17
  destroy_addr %11 : $*C                          // id: %18
  dealloc_stack %11 : $*C                         // id: %19
  destroy_value %6 : $ReferenceWritableKeyPath<C, Optional<ObjCBool>> // id: %20
  %21 = tuple ()                                  // user: %22
  return %21 : $()                                // id: %22
} // end sil function '$s4main1CPAAE3set5valueySb_tF'

In module:
sil_stage raw

import Builtin
import Swift
import SwiftShims

// main
sil [ossa] @main : $@convention(c) (Int32, UnsafeMutablePointer<Optional<UnsafeMutablePointer<Int8>>>) -> Int32 {
bb0(%0 : $Int32, %1 : $UnsafeMutablePointer<Optional<UnsafeMutablePointer<Int8>>>):

bb1:

bb2(%2 : @owned $Error):
} // end sil function 'main'

// C.set(value:)
sil [ossa] @$s4main1CPAAE3set5valueySb_tF : $@convention(method) <Self where Self : C> (Bool, @guaranteed Self) -> () {
// %0 "value"                                     // users: %9, %2
// %1 "self"                                      // users: %4, %3
bb0(%0 : $Bool, %1 : @guaranteed $Self):
  debug_value %0 : $Bool, let, name "value", argno 1 // id: %2
  debug_value %1 : $Self, let, name "self", argno 2 // id: %3
  %4 = copy_value %1 : $Self                      // user: %5
  %5 = init_existential_ref %4 : $Self : $Self, $C // user: %12
  %6 = keypath $ReferenceWritableKeyPath<C, Optional<ObjCBool>>, (objc "field1"; root $C; settable_property $ObjCBool,  id #C.field1!getter.foreign : <Self where Self : C> (Self) -> () -> ObjCBool, getter @$s4main1CP6field110ObjectiveC8ObjCBoolVvpAaB_pTK : $@convention(thin) (@in_guaranteed C) -> @out ObjCBool, setter @$s4main1CP6field110ObjectiveC8ObjCBoolVvpAaB_pTk : $@convention(thin) (@in_guaranteed ObjCBool, @in_guaranteed C) -> ()) // users: %20, %16
  %7 = metatype $@thin ObjCBool.Type              // user: %9
  // function_ref ObjCBool.init(_:)
  %8 = function_ref @$s10ObjectiveC8ObjCBoolVyACSbcfC : $@convention(method) (Bool, @thin ObjCBool.Type) -> ObjCBool // user: %9
  %9 = apply %8(%0, %7) : $@convention(method) (Bool, @thin ObjCBool.Type) -> ObjCBool // user: %10
  %10 = enum $Optional<ObjCBool>, #Optional.some!enumelt, %9 : $ObjCBool // user: %14
  %11 = alloc_stack $C                            // users: %19, %18, %16, %12
  store %5 to [init] %11 : $*C                    // id: %12
  %13 = alloc_stack $Optional<ObjCBool>           // users: %17, %16, %14
  store %10 to [trivial] %13 : $*Optional<ObjCBool> // id: %14
  // function_ref swift_setAtReferenceWritableKeyPath
  %15 = function_ref @swift_setAtReferenceWritableKeyPath : $@convention(thin) <τ_0_0, τ_0_1> (@in_guaranteed τ_0_0, @guaranteed ReferenceWritableKeyPath<τ_0_0, τ_0_1>, @in τ_0_1) -> () // user: %16
  %16 = apply %15<C, Optional<ObjCBool>>(%11, %6, %13) : $@convention(thin) <τ_0_0, τ_0_1> (@in_guaranteed τ_0_0, @guaranteed ReferenceWritableKeyPath<τ_0_0, τ_0_1>, @in τ_0_1) -> ()
  dealloc_stack %13 : $*Optional<ObjCBool>        // id: %17
  destroy_addr %11 : $*C                          // id: %18
  dealloc_stack %11 : $*C                         // id: %19
  destroy_value %6 : $ReferenceWritableKeyPath<C, Optional<ObjCBool>> // id: %20
  %21 = tuple ()                                  // user: %22
  return %21 : $()                                // id: %22
} // end sil function '$s4main1CPAAE3set5valueySb_tF'

// key path getter for C.field1 : C
sil shared [thunk] [ossa] @$s4main1CP6field110ObjectiveC8ObjCBoolVvpAaB_pTK : $@convention(thin) (@in_guaranteed C) -> @out ObjCBool {
// %0                                             // user: %10
// %1                                             // user: %3
bb0(%0 : $*ObjCBool, %1 : $*C):
  %2 = alloc_stack $C                             // users: %12, %4, %3
  copy_addr %1 to [initialization] %2 : $*C       // id: %3
  %4 = load [take] %2 : $*C                       // user: %5
  %5 = open_existential_ref %4 : $C to $@opened("368C6090-0D11-11EB-8C01-D0817AD6AF45") C // users: %11, %8, %6
  %6 = begin_borrow %5 : $@opened("368C6090-0D11-11EB-8C01-D0817AD6AF45") C // users: %9, %8, %7
  %7 = objc_method %6 : $@opened("368C6090-0D11-11EB-8C01-D0817AD6AF45") C, #C.field1!getter.foreign : <Self where Self : C> (Self) -> () -> ObjCBool, $@convention(objc_method) <τ_0_0 where τ_0_0 : C> (τ_0_0) -> ObjCBool // user: %8
  %8 = apply %7<@opened("368C6090-0D11-11EB-8C01-D0817AD6AF45") C>(%6) : $@convention(objc_method) <τ_0_0 where τ_0_0 : C> (τ_0_0) -> ObjCBool // type-defs: %5; user: %10
  end_borrow %6 : $@opened("368C6090-0D11-11EB-8C01-D0817AD6AF45") C // id: %9
  store %8 to [trivial] %0 : $*ObjCBool           // id: %10
  destroy_value %5 : $@opened("368C6090-0D11-11EB-8C01-D0817AD6AF45") C // id: %11
  dealloc_stack %2 : $*C                          // id: %12
  %13 = tuple ()                                  // user: %14
  return %13 : $()                                // id: %14
} // end sil function '$s4main1CP6field110ObjectiveC8ObjCBoolVvpAaB_pTK'

// key path setter for C.field1 : C
sil shared [thunk] [ossa] @$s4main1CP6field110ObjectiveC8ObjCBoolVvpAaB_pTk : $@convention(thin) (@in_guaranteed ObjCBool, @in_guaranteed C) -> () {
// %0                                             // user: %2
// %1                                             // user: %4
bb0(%0 : $*ObjCBool, %1 : $*C):
  %2 = load [trivial] %0 : $*ObjCBool             // user: %8
  %3 = alloc_stack $C                             // users: %10, %5, %4
  copy_addr %1 to [initialization] %3 : $*C       // id: %4
  %5 = load [take] %3 : $*C                       // user: %6
  %6 = open_existential_ref %5 : $C to $@opened("368C6090-0D11-11EB-8C01-D0817AD6AF45") C // users: %9, %8, %8, %7
  %7 = objc_method %6 : $@opened("368C6090-0D11-11EB-8C01-D0817AD6AF45") C, #C.field1!setter.foreign : <Self where Self : C> (Self) -> (ObjCBool) -> (), $@convention(objc_method) <τ_0_0 where τ_0_0 : C> (ObjCBool, τ_0_0) -> () // user: %8
  %8 = apply %7<@opened("368C6090-0D11-11EB-8C01-D0817AD6AF45") C>(%2, %6) : $@convention(objc_method) <τ_0_0 where τ_0_0 : C> (ObjCBool, τ_0_0) -> () // type-defs: %6
  destroy_value %6 : $@opened("368C6090-0D11-11EB-8C01-D0817AD6AF45") C // id: %9
  dealloc_stack %3 : $*C                          // id: %10
  %11 = tuple ()                                  // user: %12
  return %11 : $()                                // id: %12
} // end sil function '$s4main1CP6field110ObjectiveC8ObjCBoolVvpAaB_pTk'

// ObjCBool.init(_:)
sil [transparent] [serialized] @$s10ObjectiveC8ObjCBoolVyACSbcfC : $@convention(method) (Bool, @thin ObjCBool.Type) -> ObjCBool

// swift_setAtReferenceWritableKeyPath
sil @swift_setAtReferenceWritableKeyPath : $@convention(thin) <τ_0_0, τ_0_1> (@in_guaranteed τ_0_0, @guaranteed ReferenceWritableKeyPath<τ_0_0, τ_0_1>, @in τ_0_1) -> ()



PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2020-10-07-a.xctoolchain/usr/bin/swift-frontend -frontend -c -primary-file main.swift -target x86_64-apple-darwin19.6.0 -enable-objc-interop -sdk /Applications/Xcode-12.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -color-diagnostics -O -target-sdk-version 10.15.6 -module-name main -o /var/folders/br/7sr_p7cj7v549x_gtkb33df00000gp/T/main-d98d63.o
1.      Apple Swift version 5.3-dev (LLVM 240312aa7333e90, Swift 15bf0478ad7c47c)
2.      While evaluating request ASTLoweringRequest(Lowering AST to SIL for file "main.swift")
3.      While silgen emitFunction SIL function "@$s4main1CPAAE3set5valueySb_tF".
 for 'set(value:)' (at main.swift:9:12)
4.      While verifying SIL function "@$s4main1CPAAE3set5valueySb_tF".
 for 'set(value:)' (at main.swift:9:12)
0  swift-frontend           0x000000010ff02e45 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  swift-frontend           0x000000010ff020a5 llvm::sys::RunSignalHandlers() + 85
2  swift-frontend           0x000000010ff03416 SignalHandler(int) + 262
3  libsystem_platform.dylib 0x00007fff67fd15fd _sigtramp + 29
4  libsystem_platform.dylib 000000000000000000 _sigtramp + 18446603338771524128
5  libsystem_c.dylib        0x00007fff67ea7808 abort + 120
6  swift-frontend           0x000000010c3a3e03 (anonymous namespace)::SILVerifier::_require(bool, llvm::Twine const&, std::__1::function<void ()> const&) + 1075
7  swift-frontend           0x000000010c3b5eef swift::SILInstructionVisitor<(anonymous namespace)::SILVerifier, void>::visit(swift::SILInstruction*) + 61663
8  swift-frontend           0x000000010c3a5b8c (anonymous namespace)::SILVerifier::visitSILBasicBlock(swift::SILBasicBlock*) + 1628
9  swift-frontend           0x000000010c39fa21 swift::SILFunction::verify(bool) const + 8273
10 swift-frontend           0x000000010bbb9ed0 swift::Lowering::SILGenModule::postEmitFunction(swift::SILDeclRef, swift::SILFunction*) + 224
11 swift-frontend           0x000000010bbb91b0 swift::Lowering::SILGenModule::emitFunctionDefinition(swift::SILDeclRef, swift::SILFunction*) + 5808
12 swift-frontend           0x000000010bbbae24 emitOrDelayFunction(swift::Lowering::SILGenModule&, swift::SILDeclRef, bool) + 228
13 swift-frontend           0x000000010bbb7aea swift::Lowering::SILGenModule::emitFunction(swift::FuncDecl*) + 122
14 swift-frontend           0x000000010bcacf0e SILGenExtension::visitFuncDecl(swift::FuncDecl*) + 142
15 swift-frontend           0x000000010bca8b8b SILGenExtension::emitExtension(swift::ExtensionDecl*) + 59
16 swift-frontend           0x000000010bca8b45 swift::Lowering::SILGenModule::visitExtensionDecl(swift::ExtensionDecl*) + 21
17 swift-frontend           0x000000010bbbd2a7 swift::ASTLoweringRequest::evaluate(swift::Evaluator&, swift::ASTLoweringDescriptor) const + 1959
18 swift-frontend           0x000000010bc9b9a0 swift::SimpleRequest<swift::ASTLoweringRequest, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> > (swift::ASTLoweringDescriptor), (swift::RequestFlags)9>::evaluateRequest(swift::ASTLoweringRequest const&, swift::Evaluator&) + 128
19 swift-frontend           0x000000010bbc1e11 llvm::Expected<swift::ASTLoweringRequest::OutputType> swift::Evaluator::getResultUncached<swift::ASTLoweringRequest>(swift::ASTLoweringRequest const&) + 497
20 swift-frontend           0x000000010bbbe7be llvm::Expected<swift::ASTLoweringRequest::OutputType> swift::Evaluator::operator()<swift::ASTLoweringRequest, (void*)0>(swift::ASTLoweringRequest const&) + 110
21 swift-frontend           0x000000010bbbe8c3 swift::performASTLowering(swift::FileUnit&, swift::Lowering::TypeConverter&, swift::SILOptions const&) + 147
22 swift-frontend           0x000000010b845b92 performCompileStepsPostSema(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 226
23 swift-frontend           0x000000010b83df9b swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 4827
24 swift-frontend           0x000000010b7d4862 main + 866
25 libdyld.dylib            0x00007fff67dd8cc9 start + 1
26 libdyld.dylib            0x0000000000000012 start + 18446603338773590858
<unknown>:0: error: unable to execute command: Abort trap: 6
<unknown>:0: error: compile command failed due to signal 6 (use -v to see invocation)
```

**Comment by kateinoigakukun** [comment]
I confirmed it does not crash the compiler now
```
swift-driver version: 1.62.15 Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51)
Target: arm64-apple-macosx13.0
```

**Comment by AnthonyLatsis** [comment]
I believe this was fixed by #42137
